temporal-sdk/src/main/java/io/temporal/common/converter/SearchAttributesPayloadConverter.java

```diff
@@ -0,0 +1,93 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.common.converter;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.PropertyAccessor;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.google.protobuf.ByteString;
+import io.temporal.api.common.v1.Payload;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Optional;
+
+public final class SearchAttributesPayloadConverter implements PayloadConverter {
+
+  private static ObjectMapper mapper;
+  private static SearchAttributesPayloadConverter converter;
+
+  private SearchAttributesPayloadConverter() {
+    mapper = new ObjectMapper();
+    mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+    mapper.registerModule(new JavaTimeModule());
+    mapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
+  }
+
+  public static SearchAttributesPayloadConverter getInstance() {
+    if (converter == null) {
+      converter = new SearchAttributesPayloadConverter();
+    }
+    return converter;
+  }
+
+  @Override
+  public String getEncodingType() {
+    return EncodingKeys.METADATA_ENCODING_JSON_NAME;
+  }
+
+  @Override
+  public Optional<Payload> toData(Object obj) throws DataConverterException {
+    try {
+      byte[] serialized = mapper.writeValueAsBytes(obj);
+      return Optional.of(
+          Payload.newBuilder()
+              .putMetadata(EncodingKeys.METADATA_ENCODING_KEY, EncodingKeys.METADATA_ENCODING_JSON)
+              .putMetadata(
+                  EncodingKeys.METADATA_TYPE_KEY,
+                  ByteString.copyFromUtf8(
+                      // Handle NPE
+                      SearchAttributesUtil.javaTypeToEncodedType(obj.getClass()).name()))
```

Comment:
What if we get Unspecified here?

Reply:
We should discuss this: on options is to treat is a string

Reply:
Talked to Alex. Unspecified should throw an error